htlcswitch/switch_test.go

```diff
@@ -3420,6 +3423,73 @@ func TestSwitchHoldForward(t *testing.T) {
 	assertNumCircuits(t, s, 0, 0)
 
 	require.NoError(t, switchForwardInterceptor.Stop())
+
+	// Test always-on interception.
+	switchForwardInterceptor = NewInterceptableSwitch(s, true)
+	require.NoError(t, switchForwardInterceptor.Start())
+
+	// Forward a fresh packet. It is expected to be failed immediately,
+	// because there is no interceptor registered.
+	require.NoError(t, switchForwardInterceptor.ForwardPackets(
+		linkQuit, false, ogPacket,
+	))
+
+	assertOutgoingLinkReceive(t, bobChannelLink, false)
+	assertOutgoingLinkReceive(t, aliceChannelLink, true)
+	assertNumCircuits(t, s, 0, 0)
+
+	// Forward a replayed packet. It is expected to be held until the
+	// interceptor connects. To continue the test, it needs to be ran in a
+	// goroutine.
+	errChan := make(chan error)
+	go func() {
+		errChan <- switchForwardInterceptor.ForwardPackets(
+			linkQuit, true, ogPacket,
+		)
+	}()
+
+	// Assert that nothing is forward to the switch.
+	assertNumCircuits(t, s, 0, 0)
```

Comment:
To avoid nondeterministic behavior, I'd probably switch the order of `assertNumCircuits` and `assertOutgoingLinkReceive` (even though logically the order is right) since there's a chance that the above call to `ForwardPackets` will be after `assertNumCircuits`. There's a 1 second wait in `assertOutgoingLinkReceive` that makes the ordering correct (with a high chance). Alternatively we could add a forced 1 sec sleep before the asserts.